packages/components/src/custom-select-control-v2/types.ts

```diff
@@ -40,15 +33,131 @@ export type CustomSelectProps = {
 	/**
 	 * The size of the control.
 	 *
-	 * @default 'default'
+	 * @default default
 	 */
-	size?: 'default' | 'small';
+	size?: 'compact' | 'default' | 'small';
 	/**
 	 * Can be used to externally control the value of the control.
 	 */
 	value?: string | string[];
 };
 
+export type _CustomSelectProps = {
+	/**
+	 * The child elements. This should be composed of CustomSelectItem components.
+	 */
+	children: React.ReactNode;
+	/**
+	 * Used to visually hide the label. It will always be visible to screen readers.
+	 *
+	 * @default false
+	 */
+	hideLabelFromVision?: boolean;
+	/**
+	 * Label for the control.
+	 */
+	label: string;
+};
+
+export type CustomSelectProps = _CustomSelectProps &
+	Exclude< CustomSelectButtonProps, { size: 'small' } >;
+
+/**
+ * The legacy object structure for the options array.
+ */
+type Option = {
+	key: string;
+	name: string;
+	style?: {};
+	className?: string;
+	__experimentalHint?: string;
+};
+
+/**
+ * The legacy object returned from the onChange event.
+ */
+type OnChangeObject = {
+	selectedItem: Option;
+	highlightedIndex?: number;
+	isOpen?: boolean;
+};
+
+export type LegacyCustomSelectProps = {
```

Comment:
One important thing we need to do to make all these type defs usable in practice is to let the TS compiler know whether the consumer is trying to use the component with the legacy or current APIs.

For example, when a consumer does `<CustomSelectControl options={ [] }>` we want the compiler to know that the `onChange` should have the legacy signature. Similarly, if the consumer passes `children`, it should know that the `onChange` should have the new signature.

We've been doing this by adding an explicit `never` type like this, so TS knows that the presence of `children` means we are not in a legacy context:

```suggestion
export type LegacyCustomSelectProps = {
	children?: never;
```

Similarly, we would need to add `options?: never;` to `CustomSelectProps`.

When the types get complex like this, I've found it useful to add "static typing tests" like for the [Button component here](https://github.com/WordPress/gutenberg/blob/8c80c0e2be01b4868aa55727b6cfcfca0ef438ba/packages/components/src/button/test/index.tsx#L622). It verifies that the types are working as expected.

So some basic things we could check for would be (pseudocode):

```jsx
<CustomSelect
	label="foo"
	options={ [] }
	onChange={ ( _: string | string[] ) => undefined } // this should error
	onChange={ ( _: { selectedItem: unknown } ) => undefined } // this shouldn't
/>;

<CustomSelect
	label="foo"
	onChange={ ( _: string | string[] ) => undefined } // this shouldn't error
	onChange={ ( _: { selectedItem: unknown } ) => undefined } // this should
>
	foobar
</CustomSelect>;
```

Reply:
Woah, TIL about static typing tests. Very cool!